File: framework/languages/java/e2e/SpotService/Client/src/main/java/systems/zlink/e2e/spotservice/client/Scenarios/SmD13Scenario.java
package systems.zlink.e2e.spotservice.client.Scenarios;

import java.util.List;
import java.util.UUID;
import systems.zlink.e2e.spotservice.shared.Contracts;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class SmD13Scenario extends SpotServiceScenarioContext {
    private SmD13Scenario(SpotServiceScenarioContext context) {
        super(context);
    }

    public static void run(SpotServiceScenarioContext context) {
        new SmD13Scenario(context).execute();
    }

    private void execute() {
        String actorId = "actor-sm-d13-" + UUID.randomUUID().toString().replace("-", "");
        ZLinkStreamConnector connector = createStreamConnector(options().streamAEndpoint());
        Contracts.ActorProfile profile = new Contracts.ActorProfile("Heartbeat", 13, List.of("heartbeat"));
        try {
            connector.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes auth = connector
                .request(new Contracts.ActorAuthReq(actorId, profile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(actorId.equals(auth.actorId()), "SM-D13 auth actor mismatch");

            Thread.sleep(1_500);
            ensure(connector.isConnected(), "SM-D13 heartbeat-enabled stream disconnected");

            Contracts.ActorEchoRes reply = connector
                .request(new Contracts.ActorEchoReq("after-heartbeat", 13, profile))
                .metadata("actor-id", actorId)
                .submit(Contracts.ActorEchoRes.class).toCompletableFuture().join();
            ensure(actorId.equals(reply.actorId()), "SM-D13 actor reply mismatch");
            ensure("play-a".equals(reply.nodeRid()), "SM-D13 actor node mismatch");
            ensure("entry:after-heartbeat".equals(reply.value()), "SM-D13 actor value mismatch");

            waitForPlayAEvidence(List.of("StreamInbound|play-a|session|ActorEchoReq"));
            System.out.println("scenario SM-D13 passed");
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("stream heartbeat scenario interrupted", error);
        } catch (Exception error) {
            throw new IllegalStateException("stream heartbeat scenario failed", error);
        } finally {
            closeQuietly(connector);
        }

    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmD13Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.framework.kotlin.*

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.createStreamConnector
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure

internal object SmD13Scenario {
    suspend fun run() {
        val actorId = "actor-sm-d13-heartbeat"
        val profile = Contracts.ActorProfile("Heartbeat", 13, listOf("heartbeat"))
        val connector = createStreamConnector(Env.get("e2e.stream.a.endpoint"))
        try {
            connector.connect().await()
            val auth = connector
                .request(Contracts.ActorAuthReq(actorId, profile))
                .awaitReply<Contracts.ActorAuthRes>()
            ensure(auth.actorId == actorId, "SM-D13 auth actor mismatch")

            Thread.sleep(700)
            ensure(connector.isConnected, "SM-D13 heartbeat-enabled stream disconnected")

            val reply = connector
                .request(Contracts.ActorEchoReq("after-heartbeat", 13, profile))
                .awaitReply<Contracts.ActorEchoRes>()
            ensure(reply.actorId == actorId, "SM-D13 actor reply mismatch")
            ensure(reply.nodeRid == "session-a", "SM-D13 actor node mismatch")
            ensure(reply.value == "entry:after-heartbeat", "SM-D13 actor value mismatch")

            println("scenario SM-D13 passed")
        } finally {
            try {
                connector.close().await()
            } catch (_: Exception) {
            }
        }
    }
}
